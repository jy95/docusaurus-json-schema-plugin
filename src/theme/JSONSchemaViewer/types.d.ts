import type { JSONSchema as Draft_07 } from "json-schema-typed/draft-07"
import type { JSONSchema as Draft_2019_09 } from "json-schema-typed/draft-2019-09"
import type { JSONSchema as Draft_2020_12 } from "json-schema-typed/draft-2020-12"

export type JSONSchema = Draft_07 | Draft_2019_09 | Draft_2020_12

// In case I want to cover one specific file
export type JSONSchema_Draft_07 = Draft_07
export type JSONSchema_Draft_2019_09 = Draft_2019_09
export type JSONSchema_Draft_2020_12 = Draft_2020_12

// Namespace, In case I need to cast to a specific type (array / object)
import type { JSONSchema as JSONSchemaNS } from "json-schema-typed"
export { JSONSchemaNS }

export type { TypeName as TypeName } from "json-schema-typed"
export type { keywords } from "json-schema-typed"
