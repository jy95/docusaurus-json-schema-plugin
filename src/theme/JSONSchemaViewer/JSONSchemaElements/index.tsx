// Component to deal with oneOf / anyOf / allOf / not
export { SchemaComposition } from "@theme/JSONSchemaViewer/JSONSchemaElements/schemaComposition"

// Component to deal with if-then-else , dependentRequired , dependentSchemas , dependencies
export { SchemaConditional } from "@theme/JSONSchemaViewer/JSONSchemaElements/SchemaConditional"

// Component to deal with Object
export { CreateObject } from "@theme/JSONSchemaViewer/JSONSchemaElements/object"

// Component to deal with Array
export { CreateArray } from "@theme/JSONSchemaViewer/JSONSchemaElements/array"

// Component to deal with String
export { default as CreateString } from "@theme/JSONSchemaViewer/JSONSchemaElements/CreateString"

// Component to deal with Boolean
export { default as CreateBoolean } from "@theme/JSONSchemaViewer/JSONSchemaElements/CreateBoolean"

// Component to deal with Number
export { default as CreateNumber } from "@theme/JSONSchemaViewer/JSONSchemaElements/CreateNumber"

// Component to deal with Integer
export { default as CreateInteger } from "@theme/JSONSchemaViewer/JSONSchemaElements/CreateInteger"

// Component to deal with null
export { default as CreateNull } from "@theme/JSONSchemaViewer/JSONSchemaElements/CreateNull"

// Component to deal with schema defined as "true"
export { default as CreateAlwaysValid } from "@theme/JSONSchemaViewer/JSONSchemaElements/CreateAlwaysValid"

// Component to deal with schema defined as "false"
export { default as CreateAlwaysInvalid } from "@theme/JSONSchemaViewer/JSONSchemaElements/CreateAlwaysInvalid"

// Component to deal with description
export { default as CreateDescription } from "@theme/JSONSchemaViewer/JSONSchemaElements/CreateDescription"
