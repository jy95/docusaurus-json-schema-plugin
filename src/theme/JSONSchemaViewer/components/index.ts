// To avoid circular dependancies issue, let's have a single place for exports and imports

export { default as Collapsible } from "./Collapsible"

export { default as CreateNodes } from "./CreateNodes"

export { default as CreateEdge } from "./CreateEdge"

export { default as SchemaItem } from "./SchemaItem/SchemaItem"
