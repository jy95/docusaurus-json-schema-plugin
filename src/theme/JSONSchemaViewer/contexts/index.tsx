export * from "@theme/JSONSchemaViewer/contexts/jsvOptions"
