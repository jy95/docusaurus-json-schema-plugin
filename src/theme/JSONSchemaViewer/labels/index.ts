// Re-export labels
export { default as ObjectLabel } from "@theme/JSONSchemaViewer/labels/Object"
export { default as ArrayLabel } from "@theme/JSONSchemaViewer/labels/Array"
export { default as TypeLabel } from "@theme/JSONSchemaViewer/labels/Type"
export { default as IfLabel } from "@theme/JSONSchemaViewer/labels/If"
export { default as ThenLabel } from "@theme/JSONSchemaViewer/labels/Then"
export { default as ElseLabel } from "@theme/JSONSchemaViewer/labels/Else"
export { default as LoadingLabel } from "@theme/JSONSchemaViewer/labels/Loading"
export { default as ErrorOccurredLabel } from "@theme/JSONSchemaViewer/labels/ErrorOccurred"
export { default as RequiredLabel } from "@theme/JSONSchemaViewer/labels/Required/Required"
export { default as DeprecatedLabel } from "@theme/JSONSchemaViewer/labels/Deprecated/Deprecated"
export { default as ReadOnlyLabel } from "@theme/JSONSchemaViewer/labels/ReadOnly/ReadOnly"
export { default as WriteOnlyLabel } from "@theme/JSONSchemaViewer/labels/WriteOnly/WriteOnly"
export { default as TrueLabel } from "@theme/JSONSchemaViewer/labels/True"
export { default as FalseLabel } from "@theme/JSONSchemaViewer/labels/False"
export { default as StringLabel } from "@theme/JSONSchemaViewer/labels/String"
export { default as BooleanLabel } from "@theme/JSONSchemaViewer/labels/Boolean"
export { default as NumberLabel } from "@theme/JSONSchemaViewer/labels/Number"
export { default as IntegerLabel } from "@theme/JSONSchemaViewer/labels/Integer"
export { default as NullLabel } from "@theme/JSONSchemaViewer/labels/Null"
export { default as AndLabel } from "@theme/JSONSchemaViewer/labels/And"
export { default as OrLabel } from "@theme/JSONSchemaViewer/labels/Or"
export { default as NotLabel } from "@theme/JSONSchemaViewer/labels/Not"
export { default as XorLabel } from "@theme/JSONSchemaViewer/labels/Xor"
