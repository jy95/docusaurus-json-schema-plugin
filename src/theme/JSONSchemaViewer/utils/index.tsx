// To keep my mind happy, let's have a place where I can import stuff without knowing their final place in internal structure

// Re export generateFriendlyName function
export { default as GenerateFriendlyName } from "@theme/JSONSchemaViewer/utils/generateFriendlyName"

// Re export QualifierMessages function
export { default as QualifierMessages } from "@theme/JSONSchemaViewer/utils/getQualifierMessages"

// Re export detectTypes functions
export * from "@theme/JSONSchemaViewer/utils/detectTypes"
