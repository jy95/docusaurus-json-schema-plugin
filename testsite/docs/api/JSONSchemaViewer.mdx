---
title: "ðŸ”Ž @theme/JSONSchemaViewer"
description: "@theme/JSONSchemaViewer API"
---

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

## Configuration

| Property        | Type          | Required ?  | Note                                                    |
|-----------------|---------------|-------------|---------------------------------------------------------|
| schema          | JSONSchema    | Mandatory   | JSON Schema Draft-07 / Draft 2019-09 / Draft 2020-12    |
| resolverOptions | IResolveOpts  | Optional    | To resolve your $ref (by default, only inline references will be dereferenced). |
| viewerOptions   | JSVOptions    | Optional    | Options for the viewer itself. |

### Types

```tsx
// Either Draft-07 / Draft 2019-09 / Draft 2020-12
type JSONSchema = unknown
```

```tsx
type IResolveOpts = {
    // "IResolveOpts" options from @stoplightio/json-ref-resolver
    // More info on https://github.com/stoplightio/json-ref-resolver
    // https://github.com/stoplightio/json-ref-resolver/blob/master/src/types.ts
}
```

```tsx
type JSVOptions = {
   /**
   * Should we display "examples" ?
   * @default false
   */
    showExamples?: boolean
}
```

## Examples

<Tabs>
<TabItem value="JSX" label="JSX">

```tsx
import React from "react"
import Layout from "@theme/Layout"
import JSONSchemaViewer from "@theme/JSONSchemaViewer"

export default function ExamplePage(): JSX.Element {

    // You are free to fetch your schema in your own way (load local file, fetch, ...) :)
    const mySchema = {
        "type": "object",
        "properties": {
            "builtin": {
                "type": "number"
            }
        },
        "patternProperties": {
            "^S_": {
                "type": "string"
            },
            "^I_": {
                "type": "integer"
            }
        },
        "additionalProperties": {
            "type": "string"
        }
    }

    return (
        <Layout
            title={`My super JSON Schema`}
            description="Description will go into a meta tag in <head />"
        >
            <JSONSchemaViewer schema={mySchema} />
        </Layout>
    )
}
```

</TabItem>

<TabItem value="MDX" label="MDX" default>

```tsx

import CodeBlock from '@theme/CodeBlock';
// To fetch a JSON file from your static folder
import Schema from "@site/static/schemas/examples/array/tuples.json";
import JSONSchemaViewer from "@theme/JSONSchemaViewer";

# Tuples

Viewer :

<JSONSchemaViewer schema={ Schema  } />

Source :

<CodeBlock language="json">{JSON.stringify(Schema, null, 2)}</CodeBlock>

```

</TabItem>

</Tabs>