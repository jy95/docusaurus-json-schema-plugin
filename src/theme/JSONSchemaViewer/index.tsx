import React from "react"
import { Resolver } from "@stoplight/json-ref-resolver"

import { CreateNodes, Collapsible } from "./components/index"

import type { JSONSchema7 } from "json-schema"
import type { IResolverOpts } from "@stoplight/json-ref-resolver/types"

export type Props = {
  schema: JSONSchema7
  // To customize the ref resolving
  resolverOptions?: IResolverOpts
  [x: string]: any
}

type InnerViewerProperties = {
  // Thanks to @stoplight/json-ref-resolver, $ref are either :
  // 1. resolved
  // 2. unresolved (as circular stuff are not on the roadmap)
  schema: JSONSchema7
}

// Internal
function JSONSchemaInnerViewer(props: InnerViewerProperties): JSX.Element {
  const { schema } = props
  // Title of the schema, for user friendliness
  const title = schema?.title || "Schema"
  return (
    <Collapsible
      summary={<strong>{title}</strong>}
      detailsProps={{
        open: true,
      }}
    >
      <CreateNodes schema={schema} />
    </Collapsible>
  )
}

// Entry point
function JSONSchemaViewer(props: Props): JSX.Element {
  const { schema: originalSchema, resolverOptions } = props

  // Set up resolver
  // @ts-ignore
  const resolver = new Resolver(resolverOptions)

  // TODO How do I wait for async to be complete before I run that ?
  // resolver.resolve(originalSchema);

  // Simplify schema
  //const simplifiedSchema = (originalSchema?.allOf !== undefined) ? mergeAllOf(originalSchema) : originalSchema;

  return <JSONSchemaInnerViewer schema={originalSchema} />
}

export default JSONSchemaViewer
