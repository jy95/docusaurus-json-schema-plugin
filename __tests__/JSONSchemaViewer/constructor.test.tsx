import React from "react"
import { expect, test, describe } from "@jest/globals"
import { render, act } from "@testing-library/react"
import type { RenderResult } from "@testing-library/react"
import JSONSchemaViewer from "../../src/theme/JSONSchemaViewer/index"
import type { JSONSchema } from "../../src/theme/JSONSchemaViewer/types"
import CodeBlock from "../../__mocks__/@theme-original/CodeBlock"

describe("JSONSchemaViewer - constructor", () => {
  test("Overwrite default qualifierMessagesOrder value", async () => {
    const fakeSchema: JSONSchema = { type: "object", minProperties: 1 }
    let result: RenderResult | null = null

    await act(async () => {
      result = render(
        <JSONSchemaViewer
          schema={fakeSchema}
          viewerOptions={{ qualifierMessagesOrder: ["objectProperties"] }}
        />,
      )
    })

    expect(result!.asFragment()).toMatchSnapshot()
  })

  test("Overwrite default DescriptionComponent value", async () => {
    const fakeSchema2: JSONSchema = {
      type: "object",
      description: "# Hello, *world*!",
    }
    let result: RenderResult | null = null

    await act(async () => {
      result = render(
        <JSONSchemaViewer
          schema={fakeSchema2}
          viewerOptions={{
            DescriptionComponent: () => (
              <h1>
                Hello, <em>world</em>!
              </h1>
            ),
          }}
        />,
      )
    })

    expect(result!.asFragment()).toMatchSnapshot()
  })

  test("Overwrite default UnresolvedRefsComponent value", async () => {
    const fakeSchema: JSONSchema = {
      $schema: "https://json-schema.org/draft/2020-12/schema",
      $id: "https://example.com/tree",
      $dynamicAnchor: "node",
      type: "object",
      properties: {
        data: true,
        children: { type: "array", items: { $dynamicRef: "#node" } },
      },
    }
    let result: RenderResult | null = null

    await act(async () => {
      result = render(
        <JSONSchemaViewer
          schema={fakeSchema}
          viewerOptions={{
            UnresolvedRefsComponent: () => <>#node was not resolved</>,
          }}
        />,
      )
    })

    expect(result!.asFragment()).toMatchSnapshot()
  })

  test("Overwrite default ValueComponent value", async () => {
    const fakeSchema: JSONSchema = {
      $schema: "http://json-schema.org/draft-07/schema#",
      title: "CustomizationOptions",
      description: "JSON schema for customized options",
      type: "object",
      properties: {
        customField: {
          type: "string",
          description: "A customized or personalized field",
          enum: [
            "palette",
            "teddyBear",
            "tools",
            "laptop",
            "thread",
            "phone",
            "puzzle",
            "scissors",
            "hammer",
            "note",
          ],
          default: "palette",
          examples: ["tools", "note"],
        },
        customConstObject: { type: "object", const: { version: 5 } },
      },
      required: ["customField"],
      additionalProperties: false,
    }
    let result: RenderResult | null = null

    await act(async () => {
      result = render(
        <JSONSchemaViewer
          schema={fakeSchema}
          viewerOptions={{
            ValueComponent: ({ value, schema }) => {
              if (!["string", "number", "undefined"].includes(typeof value)) {
                return (
                  <CodeBlock language="json">{`${JSON.stringify(value, null, 2)}`}</CodeBlock>
                )
              }

              const component = <code>{`${value}`}</code>

              if (
                typeof schema !== "boolean" &&
                schema.default &&
                value === schema.default
              ) {
                return <strong>{component}</strong>
              }

              return component
            },
          }}
        />,
      )
    })

    expect(result!.asFragment()).toMatchSnapshot()
  })

  test("Overwrite default className value", async () => {
    const fakeSchema: JSONSchema = { type: "object", minProperties: 1 }
    let result: RenderResult | null = null

    await act(async () => {
      result = render(
        <JSONSchemaViewer schema={fakeSchema} className="jsv-custom" />,
      )
    })

    expect(result!.asFragment()).toMatchSnapshot()
  })
})
