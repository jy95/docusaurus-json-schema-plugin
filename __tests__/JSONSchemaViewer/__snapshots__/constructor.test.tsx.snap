// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JSONSchemaViewer - constructor Overwrite default DescriptionComponent value 1`] = `
<DocumentFragment>
  <details
    class="json-schema-viewer"
    open=""
  >
    <summary>
      <strong>
        Schema
      </strong>
    </summary>
    <div>
      <div>
        <strong>
          type
        </strong>
        Â :Â 
        <span
          style="opacity: 0.6;"
        >
          object
        </span>
        <div />
        <div>
          <h1>
            Hello, 
            <em>
              world
            </em>
            !
          </h1>
        </div>
      </div>
    </div>
  </details>
</DocumentFragment>
`;

exports[`JSONSchemaViewer - constructor Overwrite default UnresolvedRefsComponent value 1`] = `
<DocumentFragment>
  <details
    class="json-schema-viewer"
    open=""
  >
    <summary>
      <strong>
        Schema
      </strong>
    </summary>
    <div>
      <div>
        <strong>
          type
        </strong>
        Â :Â 
        <span
          style="opacity: 0.6;"
        >
          object
        </span>
        <ul>
          <li
            data-level="1"
            id="/properties/data"
          >
            <details>
              <summary>
                <strong>
                  data
                </strong>
                Â 
                <span
                  style="opacity: 0.6;"
                >
                  any
                </span>
              </summary>
              <div>
                <div>
                  <strong>
                    type
                  </strong>
                  Â :Â 
                  <span
                    style="opacity: 0.6;"
                  >
                    any
                  </span>
                  <div>
                    Always valid
                  </div>
                </div>
              </div>
            </details>
          </li>
          <li
            data-level="1"
            id="/properties/children"
          >
            <details>
              <summary>
                <strong>
                  children
                </strong>
                Â (
                <span
                  style="opacity: 0.6;"
                >
                  any
                </span>
                ,...)[]
              </summary>
              <div>
                <div>
                  <strong>
                    type
                  </strong>
                  Â :Â 
                  <span
                    style="opacity: 0.6;"
                  >
                    array
                  </span>
                  <ul>
                    <li
                      data-level="2"
                      id="/properties/children/items"
                    >
                      <details>
                        <summary>
                          <code>
                            items[x]
                          </code>
                          Â 
                          <span
                            style="opacity: 0.6;"
                          >
                            any
                          </span>
                        </summary>
                        <div>
                          <div>
                            <strong>
                              type
                            </strong>
                            Â :Â 
                            <span
                              style="opacity: 0.6;"
                            >
                              any
                            </span>
                            <div>
                              Always valid
                            </div>
                            <div>
                              <div>
                                #node was not resolved
                              </div>
                            </div>
                          </div>
                        </div>
                      </details>
                    </li>
                  </ul>
                  <div />
                </div>
              </div>
            </details>
          </li>
        </ul>
        <div />
      </div>
    </div>
  </details>
</DocumentFragment>
`;

exports[`JSONSchemaViewer - constructor Overwrite default ValueComponent value 1`] = `
<DocumentFragment>
  <details
    class="json-schema-viewer"
    open=""
  >
    <summary>
      <strong>
        CustomizationOptions
      </strong>
    </summary>
    <div>
      <div>
        <strong>
          type
        </strong>
        Â :Â 
        <span
          style="opacity: 0.6;"
        >
          object
        </span>
        <ul>
          <li
            data-level="1"
            id="/properties/customField"
          >
            <details>
              <summary>
                <strong>
                  customField
                </strong>
                Â 
                <span
                  style="opacity: 0.6;"
                >
                  string
                </span>
                Â 
                <strong>
                  required
                </strong>
              </summary>
              <div>
                <div>
                  <strong>
                    type
                  </strong>
                  Â :Â 
                  <span
                    style="opacity: 0.6;"
                  >
                    string
                  </span>
                  <div>
                    <div>
                      <strong>
                        Possible values :
                      </strong>
                      Â 
                      <ul>
                        <li>
                          <strong>
                            <code>
                              palette
                            </code>
                          </strong>
                        </li>
                        <li>
                          <code>
                            teddyBear
                          </code>
                        </li>
                        <li>
                          <code>
                            tools
                          </code>
                        </li>
                        <li>
                          <code>
                            laptop
                          </code>
                        </li>
                        <li>
                          <code>
                            thread
                          </code>
                        </li>
                        <li>
                          <code>
                            phone
                          </code>
                        </li>
                        <li>
                          <code>
                            puzzle
                          </code>
                        </li>
                        <li>
                          <code>
                            scissors
                          </code>
                        </li>
                        <li>
                          <code>
                            hammer
                          </code>
                        </li>
                        <li>
                          <code>
                            note
                          </code>
                        </li>
                      </ul>
                    </div>
                    <div>
                      <strong>
                        Default value :
                      </strong>
                      Â 
                      <strong>
                        <code>
                          palette
                        </code>
                      </strong>
                    </div>
                  </div>
                  <div>
                    A customized or personalized field
                  </div>
                </div>
              </div>
            </details>
          </li>
          <li
            data-level="1"
            id="/properties/customConstObject"
          >
            <details>
              <summary>
                <strong>
                  customConstObject
                </strong>
                Â 
                <span
                  style="opacity: 0.6;"
                >
                  object
                </span>
                Â 
                <strong>
                  constant
                </strong>
              </summary>
              <div>
                <div>
                  <strong>
                    type
                  </strong>
                  Â :Â 
                  <span
                    style="opacity: 0.6;"
                  >
                    object
                  </span>
                  <div>
                    <div>
                      <strong>
                        Constant value :
                      </strong>
                      Â 
                      <code>
                        {
  "version": 5
}
                      </code>
                    </div>
                  </div>
                </div>
              </div>
            </details>
          </li>
        </ul>
        <div>
          <div>
            ðŸš¨Â No extra propertie(s) are authorized in this object
          </div>
        </div>
        <div>
          JSON schema for customized options
        </div>
      </div>
    </div>
  </details>
</DocumentFragment>
`;

exports[`JSONSchemaViewer - constructor Overwrite default className value 1`] = `
<DocumentFragment>
  <details
    class="jsv-custom"
    open=""
  >
    <summary>
      <strong>
        Schema
      </strong>
    </summary>
    <div>
      <div>
        <strong>
          type
        </strong>
        Â :Â 
        <span
          style="opacity: 0.6;"
        >
          object
        </span>
        <div>
          <div>
            <strong>
              Length :
            </strong>
            Â 
            <code>
              &gt;= 1 propertie(s)
            </code>
          </div>
        </div>
      </div>
    </div>
  </details>
</DocumentFragment>
`;

exports[`JSONSchemaViewer - constructor Overwrite default qualifierMessagesOrder value 1`] = `
<DocumentFragment>
  <details
    class="json-schema-viewer"
    open=""
  >
    <summary>
      <strong>
        Schema
      </strong>
    </summary>
    <div>
      <div>
        <strong>
          type
        </strong>
        Â :Â 
        <span
          style="opacity: 0.6;"
        >
          object
        </span>
        <div>
          <div>
            <strong>
              Length :
            </strong>
            Â 
            <code>
              &gt;= 1 propertie(s)
            </code>
          </div>
        </div>
      </div>
    </div>
  </details>
</DocumentFragment>
`;
